### PR TITLE
fix(ux): get rid of duplicated tracebacks

### LINKED ISSUE #9990: docs: Exception stack traces on the "Datatypes and Datashapes" page
### Please describe the issue

On the "Datatypes and Datashapes" page, there are some exception stack traces in the example results:

https://ibis-project.org/concepts/datatypes

I did not check other documentation pages for similar exceptions; that might be worth a look, if there's a programmatic way to do it.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Based on #9811, I removed the code (`_repr_mimebundle_`) that was causing duplicated tracebacks. Instead I implemented `_repr_pretty_` (for the consoles) and `_repr_html_` for the notebook.

Closes #9990.

### CODE REVIEW

File: ibis/expr/types/core.py [hunk 1]
```diff
@@ -68,21 +50,31 @@ class Expr(Immutable, Coercible):
...
             return self._noninteractive_repr()
 
-    def __rich_console__(self, console: Console, options):
-        from rich.text import Text
+    def _repr_pretty_(self, p, cycle):
+        p.text(_capture_rich_renderable(self))
+
+    def _repr_html_(self):
+        from rich import get_console
+        from rich.jupyter import _render_segments
```
[the remark this answers is not in the document]
Comment (jcrist) [reply ibis/expr/types/core.py:70]: Assuming this is inlining a bit of what rich's `JupyterMixin` used to get us, would you mind adding a small comment here saying so so we know why we're touching private API here?

**Review by jcrist — APPROVED** [review]

**Comment by cpcloud** [comment]
There is no solution that fixes the issues we see *and* reprs tables aesthetically the same.

Here's the issue:

When both `_repr_html_` and either `__repr__` or `_repr_pretty_` are defined, both `_repr_html_` and the other (but of course not **both**!) are called. In our case, since they're both sharing the same rich console rendering code, they both raise the same exception which gets printed to the notebook **and** then raised, so it shows up twice.

## Possible solutions

### Define only `_repr_html_`

If only `_repr_html_` is defined, then it's not possible to invoke `__repr__` with an exception (and not even possible to call `_repr_pretty_` at all), so the exception appears once. Since `_repr_html_` only works in the notebook, this solution is a non-starter.

### Define only `_repr_pretty_`

If only `_repr_pretty_` is defined, then apparently `__repr__` isn't called (contrary to what's said in https://github.com/ipython/ipython/issues/9771), or least only a single traceback shows up.

The output of that however isn't nearly as nice as with `_repr_html_`.

### Define only `__repr__`

This is the same as defining only `_repr_pretty_`.

### This hack that might work: `None`-ing a failing `_repr_html_` into oblivion 🕳️ 

Reading the docs I saw this:

image

I tried catching the exception **only in `_repr_html_`**, and that seems to do the trick:

1. `__repr__` is unaffected
2. `_repr_pretty_` is unaffected
3. `_repr_html_` only renders the traceback once
4. all the things remain aesthetically the same

Ergo, hacking provides for nice things (??)

**Comment by cpcloud** [comment]
I can also bring back `_repr_mimebundle_` now I believe.

File: ibis/expr/types/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/expr/types/core.py]: Reverted back to using `_FixedTextJupyterMixin` so this private usage isn't necessary anymore.

**Comment by cpcloud** [comment]
I have something that can also improve the exception look on that page, will do in a follow-up.

**Comment by cpcloud** [comment]
Just to wrap this up, here's the current state of the art for the various interpreters/notebooks:

## Code

```python
from ibis.interactive import *

t = ibis.memtable({"x": [1, 2, 3]})
t2 = ex.penguins.fetch()

t2.island + t.x.cast(str)  # expected exception
```

## Vanilla Python REPL

```pycon
>>> from ibis.interactive import *
>>>
>>> t = ibis.memtable({"x": [1, 2, 3]})
>>> t2 = ex.penguins.fetch()
>>>
>>> t2.island + t.x.cast(str)  # expected exception
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/cloud/src/github.com/ibis/ibis/expr/types/core.py", line 83, in __repr__
    return _capture_rich_renderable(self)
  File "/home/cloud/src/github.com/ibis/ibis/expr/types/core.py", line 63, in _capture_rich_renderable
    console.print(renderable)
  File "/nix/store/kzjp79qwj05yn1jh08d7j74szsfkq08m-python3-3.10.14-env/lib/python3.10/site-packages/rich/console.py", line 1710, in print
    extend(render(renderable, render_options))
  File "/nix/store/kzjp79qwj05yn1jh08d7j74szsfkq08m-python3-3.10.14-env/lib/python3.10/site-packages/rich/console.py", line 1311, in render
    render_iterable = renderable.__rich_console__(self, _options)  # type: ignore[union-attr]
  File "/home/cloud/src/github.com/ibis/ibis/expr/types/core.py", line 106, in __rich_console__
    rich_object = to_rich(self, console_width=console_width)
  File "/home/cloud/src/github.com/ibis/ibis/expr/types/pretty.py", line 274, in to_rich
    return _to_rich_table(
  File "/home/cloud/src/github.com/ibis/ibis/expr/types/pretty.py", line 318, in _to_rich_table
    table = tablish.as_table()
  File "/home/cloud/src/github.com/ibis/ibis/expr/types/generic.py", line 1538, in as_table
    raise com.RelationError(
ibis.common.exceptions.RelationError: Cannot convert <class 'ibis.expr.types.strings.StringColumn'> expression involving multiple base table references to a projection
```

## IPython REPL

image

## Jupyter notebook

image

## Quarto page

image

## vscode

image

**Pull request merged**